README.md
# Innorules Ruletrace Converter

[![release](https://shields.io/github/v/release/silentsoft/actlist-innorules-ruletrace-converter-plugin?display_name=tag)](https://github.com/silentsoft/actlist-innorules-ruletrace-converter-plugin/releases)
[![license](https://img.shields.io/badge/license-Apache--2.0-green.svg)](https://github.com/silentsoft/actlist-innorules-ruletrace-converter-plugin/blob/master/LICENSE.txt)
[![Hits](https://hits.sh/github.com/silentsoft/actlist-innorules-ruletrace-converter-plugin.svg)](https://hits.sh/github.com/silentsoft/actlist-innorules-ruletrace-converter-plugin/)

> Drag and drop your rule trace files to convert special delimiters.

## Features
  - Convert special delimiters on rule trace `.xml`.
    - replace `STX(0x02)`, `EOT(0x04)`, `GS(0x1D)` and `RS(0x1E)` to `:`
    - replace `ETX(0x03)` and `<CR><LF>` to `;`
    - replace `<>` to `&lt;&gt;`
    - replace `&` to `&amp;`
  - Insert XML specification to first line if not exists.
  - Automatic conversion to specific directory every 5 seconds.
  - Automatic encoding detaction (`EUC-KR` or `UTF-8`)

## Packaging
```
$ mvn clean package
```

## Contributing
Pull requests are welcome. For major changes, please open an issue first to discuss what you would like to change.

Please note we have a [CODE_OF_CONDUCT](https://github.com/silentsoft/actlist-innorules-ruletrace-converter-plugin/blob/master/CODE_OF_CONDUCT.md), please follow it in all your interactions with the project.

## License
Please refer to [LICENSE](https://github.com/silentsoft/actlist-innorules-ruletrace-converter-plugin/blob/master/LICENSE.txt) and [NOTICE](https://github.com/silentsoft/actlist-innorules-ruletrace-converter-plugin/blob/master/NOTICE.md).


CHANGELOG.md
## 1.3.0 (22 Dec 2022)
  - ### Changes ###
    - Convert `STX(0x02)`, `EOT(0x04)`, `GS(0x1D)` and `RS(0x1E)` to `:`
    - Convert `ETX(0x03)` and `<CR><LF>` to `;`
    - Convert `<>` to `&lt;&gt;`

## 1.2.1 (19 Feb 2020)
  - ### Bug fixes
    - Minor things

## 1.2.0 (19 Feb 2020)
  - ### New features
    - Convert `&` character to `&amp;`
  - ### Bug fixes
    - Thread interrupt failure
    - Config persistence failure
    - Duplicates XML specification header

## 1.1.0 (18 Feb 2020)
  - ### Bug fixes
    - Encoding failure

## 1.0.0 (18 Feb 2020)
  - The first proper release

src/test/java/RuleTraceUtilTest.java
import org.junit.Assert;
import org.junit.Test;

public class RuleTraceUtilTest {
	
	@Test
	public void convertTest() {
		Assert.assertArrayEquals(String.valueOf(PluginConst.REPLACE_VALUE_STX).toCharArray(), RuleTraceUtil.convert(Character.toString(PluginConst.REPLACE_TARGET_STX)));
		Assert.assertArrayEquals(String.valueOf(PluginConst.REPLACE_VALUE_EOT).toCharArray(), RuleTraceUtil.convert(Character.toString(PluginConst.REPLACE_TARGET_EOT)));
		Assert.assertArrayEquals(String.valueOf(PluginConst.REPLACE_VALUE_GS).toCharArray(), RuleTraceUtil.convert(Character.toString(PluginConst.REPLACE_TARGET_GS)));
		Assert.assertArrayEquals(String.valueOf(PluginConst.REPLACE_VALUE_RS).toCharArray(), RuleTraceUtil.convert(Character.toString(PluginConst.REPLACE_TARGET_RS)));

		Assert.assertArrayEquals(String.valueOf(PluginConst.REPLACE_VALUE_ETX).toCharArray(), RuleTraceUtil.convert(Character.toString(PluginConst.REPLACE_TARGET_ETX)));
		Assert.assertArrayEquals(PluginConst.REPLACE_VALUE_CRLF.toCharArray(), RuleTraceUtil.convert(PluginConst.REPLACE_TARGET_CRLF));

		Assert.assertArrayEquals(PluginConst.REPLACE_VALUE_LTGT.toCharArray(), RuleTraceUtil.convert(PluginConst.REPLACE_TARGET_LTGT));
		Assert.assertArrayEquals(PluginConst.REPLACE_VALUE_AMPERSAND.toCharArray(), RuleTraceUtil.convert("&"));

		Assert.assertArrayEquals("&amp;".toCharArray(), RuleTraceUtil.convert("&amp;"));
		Assert.assertArrayEquals("&apos;".toCharArray(), RuleTraceUtil.convert("&apos;"));
		Assert.assertArrayEquals("&lt;".toCharArray(), RuleTraceUtil.convert("&lt;"));
		Assert.assertArrayEquals("&gt;".toCharArray(), RuleTraceUtil.convert("&gt;"));
		Assert.assertArrayEquals("&quot;".toCharArray(), RuleTraceUtil.convert("&quot;"));

		Assert.assertArrayEquals("A&lt;&gt;B".toCharArray(), RuleTraceUtil.convert("A<>B"));
		Assert.assertArrayEquals("A&lt;&gt;B".toCharArray(), RuleTraceUtil.convert("A&lt;&gt;B"));
		Assert.assertArrayEquals("A&lt;&gt;B A&lt;&gt;B".toCharArray(), RuleTraceUtil.convert("A<>B A&lt;&gt;B"));
		Assert.assertArrayEquals("A&lt;&gt;B A&lt;&gt;B".toCharArray(), RuleTraceUtil.convert("A&lt;&gt;B A<>B"));

		Assert.assertArrayEquals("A&amp;B".toCharArray(), RuleTraceUtil.convert("A&B"));
		Assert.assertArrayEquals("A&amp;B".toCharArray(), RuleTraceUtil.convert("A&amp;B"));
		Assert.assertArrayEquals("A&amp;B A&amp;B".toCharArray(), RuleTraceUtil.convert("A&B A&amp;B"));
		Assert.assertArrayEquals("A&amp;B A&amp;B".toCharArray(), RuleTraceUtil.convert("A&amp;B A&B"));
	}
	
}


src/test/java/org/silentsoft/oss/NoticeFileTest.java
package org.silentsoft.oss;

import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Paths;

import org.junit.Assert;
import org.junit.Test;
import org.silentsoft.oss.license.ApacheLicense2_0;

public class NoticeFileTest {
	
	private static final License APACHE_LICENSE_2_0 = new ApacheLicense2_0();
	
	@Test
	public void noticeFileTest() throws Exception {
		String markdown = generateInnorulesRuletraceConverterNoticeMarkdown();
		
		System.out.println("--------START OF THE NOTICE FILE--------");
		System.out.println(markdown);
		System.out.println("---------END OF THE NOTICE FILE---------");
		
		Assert.assertEquals(markdown, readFile());
	}
	
	private String generateInnorulesRuletraceConverterNoticeMarkdown() {
		return NoticeFileGenerator.newInstance("Innorules Ruletrace Converter", "silentsoft.org")
			.addText("This product includes software developed by silentsoft.org.")
			.addText("This product includes software developed by The Apache Software Foundation (http://www.apache.org/).")
			.addLibrary("guessencoding", "1.4", "https://github.com/codehaus/guessencoding", APACHE_LICENSE_2_0)
			.generate();
	}
	
	private String readFile() throws Exception {
		return String.join("\r\n", Files.readAllLines(Paths.get(System.getProperty("user.dir"), "NOTICE.md"), StandardCharsets.UTF_8));
	}

}


update-check.js
//====================================================================================================================
//             This update-check.js calculates simple Semantic version which is configured like 'x.y.z'
//             the alphabetical character calculation like 'alpha' or 'beta' is excluded on this script.
//====================================================================================================================
var latestVersion = '1.3.0';		// change this value to what you want.
var jar = 'https://github.com/your-github-account/your-plugin-name/releases/download/v1.3.0/your-plugin-1.3.0.jar';
//var requiredActlist   = '1.6.0';	// fill this value if necessary.
var url = 'https://github.com/your-github-account/your-plugin-name/releases'; // change this value to what you want.
//var killSwitchUntil   = '1.1.0';	// fill this value if necessary.
//var endOfServiceUntil = '1.2.0';	// fill this value if necessary.
//====================================================================================================================


/* IIFE (Immediately Invoked Function Expression) */
(function() {
	return {
		'available': available(),
		'jar': jar,
		'requiredActlist': requiredActlist(),
		'url': url,
		'endOfService': endOfService(),
		'killSwitch': killSwitch()
	};
}());


/* The plugin's version will be invoked as 'version' variable from Actlist engine. */
var version = version;

function available() {
	return compare(latestVersion, version) > 0;
}

function requiredActlist() {
	if (typeof requiredActlist == 'undefiend') {
		return '';
	}
	
	return requiredActlist;
}

function endOfService() {
	if (typeof endOfServiceUntil == 'undefined') {
		return false;
	}
	
	return compare(version, endOfServiceUntil) < 0;
}

function killSwitch() {
	if (typeof killSwitchUntil == 'undefined') {
		return 'off';
	}
	
	return (compare(version, killSwitchUntil) < 0) ? 'on' : 'off';
}

function compare(latestVersion, version) {
	if (latestVersion == version) {
		return 0;
	}
	
	var _o1 = latestVersion.split('.');
	var _o2 = version.split('.');
	
	var o1Major = parseInt(_o1[0]);
	var o2Major = parseInt(_o2[0]);
	
	var majorCompare = o1Major.compareTo(o2Major);
	if (majorCompare == 0) {
		var o1Minor = parseInt(_o1[1]);
		var o2Minor = parseInt(_o2[1]);
		
		var minorCompare = o1Minor.compareTo(o2Minor);
		if (minorCompare == 0) {
			var o1Patch = parseInt(_o1[2]);
			var o2Patch = parseInt(_o2[2]);
			
			return o1Patch.compareTo(o2Patch);
		} else {
			return minorCompare;
		}
	}
	
	return majorCompare;
}

src/main/java/Plugin.java


import java.io.File;
import java.net.URI;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.TimeUnit;
import java.util.stream.Collectors;

import org.silentsoft.actlist.plugin.ActlistPlugin;

import com.jfoenix.controls.JFXToggleButton;
import com.sun.javafx.stage.StageHelper;

import javafx.application.Platform;
import javafx.collections.ObservableList;
import javafx.fxml.FXML;
import javafx.scene.control.Label;
import javafx.scene.control.ListView;
import javafx.scene.input.Dragboard;
import javafx.scene.input.TransferMode;
import javafx.stage.DirectoryChooser;
import javafx.stage.Stage;
import javafx.stage.Window;

public class Plugin extends ActlistPlugin {
	
	@FXML
	private JFXToggleButton automaticConvertToggleButton;
	
	@FXML
	private Label automaticConvertPathLabel;
	
	@FXML
	private ListView<RuleTraceFile> ruleTraceFileListView;
	
	private Thread directoryObserverThread = null;
	
	private List<String> convertedTraces = new ArrayList<>();
	
	public static void main(String[] args) {
		debug();
	}
	
	public Plugin() throws Exception {
		super("InnoRules Ruletrace Converter");
		
		setPluginVersion("1.3.0");
		setPluginAuthor("silentsoft.org", URI.create("https://github.com/silentsoft/actlist-innorules-ruletrace-converter-plugin"));
		setPluginUpdateCheckURI(URI.create("http://actlist.silentsoft.org/api/plugin/cc78cf69/update/check"));
		setPluginDescription(URI.create("https://github.com/silentsoft/actlist-innorules-ruletrace-converter-plugin/blob/master/README.md"));
		setPluginChangeLog(URI.create("https://github.com/silentsoft/actlist-innorules-ruletrace-converter-plugin/blob/master/CHANGELOG.md"));
		setPluginLicense(URI.create("https://github.com/silentsoft/actlist-innorules-ruletrace-converter-plugin/blob/master/NOTICE.md"));
		
		setMinimumCompatibleVersion(1, 2, 6);
	}

	@Override
	protected void initialize() throws Exception {
		automaticConvertToggleButton.selectedProperty().addListener((observable, oldValue, newValue) -> {
			try {
				if (newValue) {
					boolean shouldStartDirectoryObserverThread = true;
					
					String automaticConvertPath = getConfig(PluginConst.CONFIG_KEY_AUTOMATIC_CONVERT_PATH);
					if (automaticConvertPath == null || "".equals(automaticConvertPath.trim()) || Files.notExists(Paths.get(automaticConvertPath)) || Files.isDirectory(Paths.get(automaticConvertPath)) == false) {
						File directory = changeAutomaticConvertPath();
						if (directory == null) {
							shouldStartDirectoryObserverThread = false;
							
							automaticConvertToggleButton.setSelected(false);
						}
					}
					
					if (shouldStartDirectoryObserverThread) {
						putConfig(PluginConst.CONFIG_KEY_AUTOMATIC_CONVERT_ENABLED, true);
						
						automaticConvertPathLabel.setVisible(true);
						automaticConvertPathLabel.setText(getConfig(PluginConst.CONFIG_KEY_AUTOMATIC_CONVERT_PATH));
						
						startDirectoryObserverThread();
					}
				} else {
					putConfig(PluginConst.CONFIG_KEY_AUTOMATIC_CONVERT_ENABLED, false);
					
					stopDirectoryObserverThread();
					
					automaticConvertPathLabel.setVisible(false);
				}
			} catch (Exception e) {
				e.printStackTrace();
			}
		});
		
		automaticConvertPathLabel.setOnMouseReleased(mouseEvent -> changeAutomaticConvertPath());
		
		ruleTraceFileListView.setPlaceholder(new Label("Drag and Drop ruletrace files to here."));
		
		ruleTraceFileListView.setOnDragOver(dragEvent -> {
			Dragboard dragboard = dragEvent.getDragboard();
			if (dragboard.hasFiles()) {
				boolean containsXmlFilesOnly = true;
				
				List<File> files = dragboard.getFiles();
				for (File file : files) {
					if (file.getName().toLowerCase().endsWith(".xml") == false) {
						containsXmlFilesOnly = false;
						break;
					}
				}
				
				if (containsXmlFilesOnly) {
					dragEvent.acceptTransferModes(TransferMode.LINK);
				}
			}
			
			dragEvent.consume();
		});
		ruleTraceFileListView.setOnDragDropped(dragEvent -> {
			convertRuleTrace(dragEvent.getDragboard().getFiles());
		});
	}
	
	@Override
	public void pluginActivated() throws Exception {
		loadConfig();
		
		{
			String automaticConvertPath = getConfig(PluginConst.CONFIG_KEY_AUTOMATIC_CONVERT_PATH);
			boolean automaticConvertEnabled = getConfig(PluginConst.CONFIG_KEY_AUTOMATIC_CONVERT_ENABLED);
			
			automaticConvertPathLabel.setVisible(automaticConvertEnabled);
			automaticConvertPathLabel.setText(automaticConvertPath);
			automaticConvertToggleButton.setSelected(automaticConvertEnabled);
		}
		
		if (automaticConvertToggleButton.isSelected()) {
			startDirectoryObserverThread();
		}
	}
	
	private void loadConfig() throws Exception {
		if (getConfig(PluginConst.CONFIG_KEY_AUTOMATIC_CONVERT_ENABLED) == null) {
			putConfig(PluginConst.CONFIG_KEY_AUTOMATIC_CONVERT_ENABLED, false);
		}
		
		if (getConfig(PluginConst.CONFIG_KEY_AUTOMATIC_CONVERT_PATH) == null) {
			Path innorulesTracePath = Paths.get(PluginConst.DEFAULT_AUTOMATIC_CONVERT_PATH);
			if (Files.exists(innorulesTracePath) && Files.isDirectory(innorulesTracePath)) {
				putConfig(PluginConst.CONFIG_KEY_AUTOMATIC_CONVERT_PATH, PluginConst.DEFAULT_AUTOMATIC_CONVERT_PATH);
				putConfig(PluginConst.CONFIG_KEY_AUTOMATIC_CONVERT_ENABLED, true);
			}
		}
	}

	@Override
	public void pluginDeactivated() throws Exception {
		stopDirectoryObserverThread();
	}
	
	private void startDirectoryObserverThread() {
		if (directoryObserverThread != null) {
			directoryObserverThread.interrupt();
		}

		directoryObserverThread = null;
		directoryObserverThread = new Thread(() -> {
			try {
				while (true) {
					try {
						Path automaticConvertPath = Paths.get((String) getConfig(PluginConst.CONFIG_KEY_AUTOMATIC_CONVERT_PATH));
						if (Files.exists(automaticConvertPath)) {
							List<Path> notConvertedRuleTraces = Files.walk(automaticConvertPath, 1)
									.filter(path -> path.toString().toLowerCase().endsWith(".xml") && convertedTraces.contains(path.toString()) == false)
									.collect(Collectors.toList());
							notConvertedRuleTraces.forEach(path -> {
								convertRuleTrace(path.toFile());
								
								convertedTraces.add(path.toString());
							});
						}
					} catch (Exception e) {
						e.printStackTrace();
					}
					
					Thread.sleep(TimeUnit.SECONDS.toMillis(5));
				}
			} catch (InterruptedException e) {
				// expected exception.
			}
		});
		directoryObserverThread.start();
	}
	
	private void stopDirectoryObserverThread() {
		if (directoryObserverThread != null) {
			directoryObserverThread.interrupt();
			directoryObserverThread = null;
		}
	}
	
	private File changeAutomaticConvertPath() {
		File directory = null;
		
		try {
			DirectoryChooser directoryChooser = new DirectoryChooser();
			Path innorulesPath = Paths.get(PluginConst.DEFAULT_AUTOMATIC_CONVERT_PATH);
			if (Files.exists(innorulesPath)) {
				directoryChooser.setInitialDirectory(innorulesPath.toFile());
			}
			directoryChooser.setTitle("Select trace directory to automatic convert");
			
			Window window = null;
			{
				ObservableList<Stage> stages = StageHelper.getStages();
				if (stages.isEmpty() == false) {
					window = stages.get(0);
				}
			}
			directory = directoryChooser.showDialog(window);
			if (directory != null) {
				putConfig(PluginConst.CONFIG_KEY_AUTOMATIC_CONVERT_PATH, directory.getAbsolutePath());
				putConfig(PluginConst.CONFIG_KEY_AUTOMATIC_CONVERT_ENABLED, true);
				
				automaticConvertPathLabel.setText(directory.getAbsolutePath());
			}
		} catch (Exception e) {
			e.printStackTrace();
		}
		
		return directory;
	}
	
	private void convertRuleTrace(List<File> files) {
		if (files == null) {
			return;
		}
		
		for (File file : files) {
			convertRuleTrace(file);
		}
	}
	
	private void convertRuleTrace(File file) {
		if (file == null || Files.notExists(file.toPath())) {
			return;
		}
		
		RuleTraceFile ruleTraceFile = new RuleTraceFile(file);
		Platform.runLater(() -> {
			ruleTraceFileListView.getItems().add(0, ruleTraceFile);
		});
		ruleTraceFile.convert();
	}
	
}


src/main/java/RuleTraceFile.java
import java.io.BufferedReader;
import java.io.File;
import java.io.FileInputStream;
import java.io.FileNotFoundException;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStreamReader;
import java.nio.charset.Charset;
import java.nio.charset.StandardCharsets;
import java.util.concurrent.atomic.AtomicBoolean;

import org.apache.commons.io.IOUtils;

import com.glaforge.i18n.io.CharsetToolkit;

import javafx.application.Platform;
import javafx.scene.control.Label;
import javafx.scene.control.ProgressIndicator;
import javafx.scene.layout.HBox;
import javafx.scene.paint.Paint;
import javafx.scene.shape.SVGPath;

public class RuleTraceFile extends HBox {
	
	private File file;
	
	private HBox statusBox;
	
	public RuleTraceFile(File file) {
		this.file = file;
		
		composite();
	}
	
	private void composite() {
		statusBox = new HBox();
		markStatusAsInProgress();
		
		setSpacing(5.0);
		getChildren().addAll(statusBox, new Label(String.format("%s (%s)", file.getName(), file.getParentFile().getPath())));
	}
	
	protected void convert() {
		new Thread(() -> {
			FileInputStream fileInputStream = null;
			InputStreamReader inputStreamReader = null;
			BufferedReader bufferedReader = null;
			FileOutputStream fileOutputStream = null;
			
			AtomicBoolean hasBeenSucceed = new AtomicBoolean(false);
			try {
				StringBuilder ruleXml = new StringBuilder();
				
				Charset encoding = guessEncoding();
				
				try {
					fileInputStream = new FileInputStream(file);
					inputStreamReader = new InputStreamReader(fileInputStream, encoding);
					bufferedReader = new BufferedReader(inputStreamReader);
					
					boolean isFirstContent = true;
					String content;
					while ((content = bufferedReader.readLine()) != null) {
						if (isFirstContent) {
							if (content.length() == 0 || "".equals(content.trim())) {
								continue;
							}
							
							if (content.trim().startsWith("<?xml version=") == false) {
								content = String.format("<?xml version=\"1.0\" encoding=\"%s\"?>\r\n%s", encoding, content);
							}
							
							isFirstContent = false;
						}
						
						ruleXml.append(RuleTraceUtil.convert(content));
						ruleXml.append("\r\n");
					}
				} catch (Exception e) {
					throw e;
				} finally {
					IOUtils.closeQuietly(fileInputStream);
					IOUtils.closeQuietly(inputStreamReader);
					IOUtils.closeQuietly(bufferedReader);
				}
				
				fileOutputStream = new FileOutputStream(file);
				IOUtils.write(ruleXml, fileOutputStream, encoding);
				
				hasBeenSucceed.set(true);
			} catch (Exception e) {
				e.printStackTrace();
			} finally {
				IOUtils.closeQuietly(fileOutputStream);
				
				Platform.runLater(() -> {
					if (hasBeenSucceed.get()) {
						markStatusAsSucceed();
					} else {
						markStatusAsFailed();
					}
				});
			}
		}).start();
	}
	
	private Charset guessEncoding() throws FileNotFoundException, IOException {
		Charset charset = CharsetToolkit.guessEncoding(file, 4096, Charset.forName("EUC-KR"));
		return Charset.forName("EUC-KR").equals(charset) ? charset : StandardCharsets.UTF_8;
	}
	
	private void markStatusAsInProgress() {
		ProgressIndicator status = new ProgressIndicator(ProgressIndicator.INDETERMINATE_PROGRESS);
		status.setPrefSize(16, 16);
		
		statusBox.getChildren().setAll(status);
	}
	
	private void markStatusAsSucceed() {
		SVGPath status = new SVGPath();
		status.setFill(Paint.valueOf("#59bf53"));
		status.setContent("M9 16.2L4.8 12l-1.4 1.4L9 19 21 7l-1.4-1.4L9 16.2z");
		
		statusBox.getChildren().setAll(status);
	}
	
	private void markStatusAsFailed() {
		SVGPath status = new SVGPath();
		status.setFill(Paint.valueOf("#da4242"));
		status.setContent("M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z");
		
		statusBox.getChildren().setAll(status);
	}
	
}


src/main/java/RuleTraceUtil.java
public class RuleTraceUtil {
	
	public static char[] convert(String content) {
		content = content.replaceAll(PluginConst.REPLACE_TARGET_LTGT, PluginConst.REPLACE_VALUE_LTGT);
		content = content.replaceAll(PluginConst.REPLACE_TARGET_CRLF, PluginConst.REPLACE_VALUE_CRLF);
		content = content.replaceAll(PluginConst.REPLACE_TARGET_AMPERSAND, PluginConst.REPLACE_VALUE_AMPERSAND);
		
		char[] charArray = content.toCharArray();
		for (int i=0, j=charArray.length; i<j; i++) {
			if (charArray[i] == PluginConst.REPLACE_TARGET_STX) {
				charArray[i] = PluginConst.REPLACE_VALUE_STX;
			}

			if (charArray[i] == PluginConst.REPLACE_TARGET_EOT) {
				charArray[i] = PluginConst.REPLACE_VALUE_EOT;
			}

			if (charArray[i] == PluginConst.REPLACE_TARGET_GS) {
				charArray[i] = PluginConst.REPLACE_VALUE_GS;
			}

			if (charArray[i] == PluginConst.REPLACE_TARGET_RS) {
				charArray[i] = PluginConst.REPLACE_VALUE_RS;
			}

			if (charArray[i] == PluginConst.REPLACE_TARGET_ETX) {
				charArray[i] = PluginConst.REPLACE_VALUE_ETX;
			}
		}
		
		return charArray;
	}

}


src/main/java/PluginConst.java
public class PluginConst {
	
	public static final char REPLACE_TARGET_STX = (char) 0x02;
	public static final char REPLACE_VALUE_STX = ':';

	public static final char REPLACE_TARGET_EOT = (char) 0x04;
	public static final char REPLACE_VALUE_EOT = ':';

	public static final char REPLACE_TARGET_GS = (char) 0x1D;
	public static final char REPLACE_VALUE_GS = ':';

	public static final char REPLACE_TARGET_RS = (char) 0x1E;
	public static final char REPLACE_VALUE_RS = ':';

	public static final char REPLACE_TARGET_ETX = (char) 0x03;
	public static final char REPLACE_VALUE_ETX = ';';

	public static final String REPLACE_TARGET_CRLF = "<CR><LF>";
	public static final String REPLACE_VALUE_CRLF = ";";

	public static final String REPLACE_TARGET_LTGT = "<>";
	public static final String REPLACE_VALUE_LTGT = "&lt;&gt;";

	public static final String REPLACE_TARGET_AMPERSAND = "(?!&amp;)(?!&apos;)(?!&lt;)(?!&gt;)(?!&quot;)[&]";
	public static final String REPLACE_VALUE_AMPERSAND = "&amp;";
	
	public static final String DEFAULT_AUTOMATIC_CONVERT_PATH = "C:\\logs\\innorules\\trace";
	
	public static final String CONFIG_KEY_AUTOMATIC_CONVERT_ENABLED = "AUTOMATIC_CONVERT_ENABLED";
	public static final String CONFIG_KEY_AUTOMATIC_CONVERT_PATH = "AUTOMATIC_CONVERT_PATH";

}
